controlplane/src/core/repositories/BillingRepository.ts

```diff
@@ -0,0 +1,239 @@
+import { and, asc, eq, not } from 'drizzle-orm';
+import Stripe from 'stripe';
+import type { DB } from '../../db/index.js';
+import { organizationBilling, billingSubscriptions, billingPlans } from '../../db/schema.js';
+
+import { BillingPlanDTO } from '../../types/index.js';
+import { toISODateTime } from '../webhooks/utils.js';
+
+export const defaultPlan = process.env.DEFAULT_PLAN;
+
+/**
+ * Repository for billing related operations.
+ */
+export class BillingRepository {
+  public stripe: Stripe;
+
+  constructor(private db: DB) {
+    this.stripe = new Stripe(process.env.STRIPE_SECRET_KEY ?? '', {
+      apiVersion: '2023-10-16',
+      appInfo: {
+        name: 'WunderGraph Cosmo',
+      },
+    });
+  }
+
+  private createOrRetrieveCustomerId = async ({ id, email }: { id: string; email?: string }) => {
+    const billing = await this.db.query.organizationBilling.findFirst({
+      where: eq(organizationBilling.organizationId, id),
+      columns: {
+        id: true,
+        email: true,
+        plan: true,
+        stripeCustomerId: true,
+      },
+    });
+
+    if (billing?.stripeCustomerId) {
+      return billing.stripeCustomerId;
+    }
+
+    const customer = await this.stripe.customers.create({
+      metadata: {
+        cosmoOrganizationId: id,
+      },
+      email, // this can be undefined, we'll update it after upgrading the plan
+    });
+
+    await this.db
+      .insert(organizationBilling)
+      .values({
+        organizationId: id,
+        stripeCustomerId: customer.id,
+        email,
+      })
+      .onConflictDoUpdate({
+        target: organizationBilling.id,
+        set: {
+          organizationId: id,
+          stripeCustomerId: customer.id,
+          email,
+        },
+      });
+
+    return customer.id;
+  };
+
+  public async listPlans(): Promise<BillingPlanDTO[]> {
+    const plans = await this.db.query.billingPlans.findMany({
+      where: not(eq(billingPlans.active, false)),
+      orderBy: [asc(billingPlans.weight)],
+    });
+
+    return plans.map(({ active, features, ...plan }) => {
+      return {
+        ...plan,
```

Comment:
Can we be sure here that we don't accidentally expose all db fields to the user?

Reply:
There's nothing private in there at the moment, but could do a select on there to make sure for any future changes :)

Reply:
thanks